File: src/main/java/com/plugin/frege/linemarker/related/FregeRelatedItemLineMarkerProvider.java
package com.plugin.frege.linemarker.related;

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo;
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public class FregeRelatedItemLineMarkerProvider extends RelatedItemLineMarkerProvider {
    FregeFunctionBindingToAnnotationLineMarker functionBindingToAnnotation = new FregeFunctionBindingToAnnotationLineMarker();
    FregeNativeFunctionToDelegatedMemberLineMarker nativeFunctionToDelegatedMember = new FregeNativeFunctionToDelegatedMemberLineMarker();

    @Override
    protected void collectNavigationMarkers(@NotNull PsiElement element, @NotNull Collection<? super RelatedItemLineMarkerInfo<?>> result) {
        functionBindingToAnnotation.addMarker(element, result);
        nativeFunctionToDelegatedMember.addMarker(element, result);
    }
}

File: src/main/kotlin/com/plugin/frege/linemarker/related/FregeRelatedItemLineMarkerProvider.kt
package com.plugin.frege.linemarker.related

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider
import com.intellij.psi.PsiElement

class FregeRelatedItemLineMarkerProvider : RelatedItemLineMarkerProvider() {
    private val functionBindingToAnnotation = FregeFunctionBindingToAnnotationLineMarker()
    private val nativeFunctionToDelegatedMember = FregeNativeFunctionToDelegatedMemberLineMarker()
    private val instanceMethodToClass = FregeInstanceMethodToClassLineMarker()
    private val classAnnotationToInstance = FregeClassAnnotationToInstanceLineMarker()
    private val classBindingToInstance = FregeClassBindingToInstanceLineMarker()

    override fun collectNavigationMarkers(
        element: PsiElement,
        result: MutableCollection<in RelatedItemLineMarkerInfo<*>>
    ) {
        functionBindingToAnnotation.addMarker(element, result)
        nativeFunctionToDelegatedMember.addMarker(element, result)
        instanceMethodToClass.addMarker(element, result)
        classAnnotationToInstance.addMarker(element, result)
        classBindingToInstance.addMarker(element, result)
    }
}
